### ISSUE TITLE
How to setup feedback related to a specific bot?

### ISSUE THREAD

**Issue by jdd-software** [SOBotics/Higgs#28]
I would like to setup these possibile feedback on Hydrant (HD).

 - fn = False negative (reported stuff)
 - tp = True positive
 - fp = False positive 
 - nc = Non constructive (No-longer needed flag)
 - sk = Skip comment it's hard to classify 

Is it enough that I send'em once via `Bot/BotRegisterFeedbackTypesPost`, how do I select icon. Maybe a bot setup web page would be better?

**Comment by Fortunate-MAN** [comment]
@jdd-software [Natty is wrong too](https://chat.stackoverflow.com/transcript/message/40862157#40862157). I believe that it'll be best if correct feedbacks are introduced into new bots, instead of introducing wrong ones because Natty got it wrong.

**Comment by jdd-software** [comment]
@Fortunate-MAN so you are saying that Tuna is wrong O_O, That needs to be a true negative....

**Comment by rjrudman** [comment]
FWIW, I agree with using the correct terminology here. Natty is known to be wrong, but it's fixable. I'd rather start out correctly, rather than having to go and change it retroactively. 

In any case, for the actual issue; I'm planning on adding a dashboard to configure feedback for bots. In the meantime, yes, you'll need to use `Bot/BotRegisterFeedbackTypesPost`. The 'Icon' field will accept a unicode character as the icon. For example, a tick or a cross

**Comment by Jinx88909** [comment]
True negative sounds right to me. It's being manually report so someone thinks this is "true" (like "true positive"). It wasn't reported automatically so it's "negative" hence "true negative".

**Comment by jdd-software** [comment]
Let's push in some machine learning definitions and decide which is correct

https://developers.google.com/machine-learning/crash-course/classification/true-false-positive-negative

> A true positive is an outcome where the model correctly predicts the positive class. Similarly, a true negative is an outcome where the model correctly predicts the negative class.

> A false positive is an outcome where the model incorrectly predicts the positive class. And a false negative is an outcome where the model incorrectly predicts the negative class. 

hmm to me it's seems like the model incorrectly predicts negative class (it think the answer/comment is ok, but it is not).

Which would support the idea that we should call it **false negative**

**Comment by Bhargav-Rao** [comment]
No, the correct term is false negative. 

Tp,  Fp, Tn and Fn are all a part of a statistical measures for classifiers. They are used to measure specificity and sensitivity of a a particular classifier. 

Here the classifier is Heat Detector. In the following image let red be the rude comments and green be the proper comments. The dots on the right side are the ones which Heat Detector classified and the ones on the left are the ones which it didn't 

![](https://i.stack.imgur.com/pyoo1.png)

When a classifier classifies any point, it is considered positive. 
When a classifier does not classify a point, it is considered negative. 

So now, the right hand side is considered as positive and the left hand side as negative. 

![](https://i.stack.imgur.com/FEwuB.png) 

When the prediction is correct, it is classified as True. When the prediction is incorrect, it is classified as False. 

A manual report: 

 - is not classified by HD : hence it is negative
 - is incorrectly predicted as "not heat": hence it is false 

Therefore manual reports are false positives

**Comment by jdd-software** [comment]
@rjrudman returning to original question :) sending from bot I would do something like this

```
FeedbackType fb1 = new FeedbackType();
fb1.setName(FeedBack.TP.toString()); //tp
fb1.setColour("green");
fb1.setIcon("&#10003;");
fb1.setIsActionable(true);
fb1.setRequiredActions(0);

```
1. color you do I need to send green, red etc or is also hex values ok like  `#F35632`?
2. Icon [example](https://www.fileformat.info/info/unicode/char/2713/index.htm) which is the best format?
3. isActionable??, what is this what does it do?
4. setRequieredActions??, , what is this what does it do?

**Comment by rjrudman** [comment]
I've removed the API endpoint for bots to manage setting up feedback. As a replacement, we now have a web view to manage them, which will be only visible to admins and bot owners. Here's yours: http://higgs.sobotics.org:5555/admin/bot/feedbackTypes/1

The bot endpoint was badly implemented, and caused https://github.com/SOBotics/Higgs/issues/29. If you still want to programatically setup the feedback, see `admin/ViewBotFeedbackTypes` and `admin/EditBotFeedbackTypes`. 

`SetRequiredActions` has been removed. It originally was supposed to restrict feedback to users with a particular role. May implement this in the future, but doesn't look like it'll be useful.

`IsActionable` doesn't do anything yet. The idea behind it, is that we want Higgs to be able to figure out which reports have been 'acted on', so we can build a review queue. Setting `IsActionable` to false excludes the feedback from the calculation.
